execution/block-partitioner/src/v2/mod.rs

```diff
@@ -0,0 +1,719 @@
+
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["pre_partition_uniform"])
+            .start_timer();
+        let remaining_txn_idxs = uniform_partition(num_txns, num_executor_shards);
+        let mut start_txns: Vec<OriginalTxnIdx> = vec![0; num_executor_shards];
+        for shard_id in 1..num_executor_shards {
+            start_txns[shard_id] =
+                start_txns[shard_id - 1] + remaining_txn_idxs[shard_id - 1].len();
+        }
+        let duration = timer.stop_and_record();
+        info!("pre_partition_uniform={duration}");
+
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["multi_rounds"])
+            .start_timer();
+        let (finalized_txn_matrix, start_index_matrix, new_idxs) = self.multi_rounds(
+            num_txns,
+            num_executor_shards,
+            &rsets,
+            &wsets,
+            &senders,
+            &trackers,
+            &start_txns,
+            remaining_txn_idxs,
+        );
+        let duration = timer.stop_and_record();
+        info!("multi_rounds={duration}");
+
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["add_edges"])
+            .start_timer();
+        let ret = self.add_edges(
+            txns,
+            &rsets,
+            &wsets,
+            &finalized_txn_matrix,
+            &start_index_matrix,
+            &new_idxs,
+            &trackers,
+        );
+        let duration = timer.stop_and_record();
+        info!("add_edges={duration}");
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["drop"])
+            .start_timer();
+        self.thread_pool.spawn(move || {
+            drop(sender_idx_table);
+            drop(senders);
+            drop(wsets);
+            drop(rsets);
+            drop(trackers);
+            drop(key_idx_table);
+            drop(start_txns);
+            drop(finalized_txn_matrix);
+            drop(start_index_matrix);
+            drop(new_idxs);
+        });
+        let duration = timer.stop_and_record();
+        info!("drop={duration}");
+        ret
+    }
+}
+
+#[test]
+fn test_partitioner_v2_correctness() {
+    let block_generator = P2PBlockGenerator::new(100);
+    let partitioner = PartitionerV2::new(8, 4, 10, 64);
+    let mut rng = thread_rng();
+    for _run_id in 0..100 {
+        let block_size = 10_u64.pow(rng.gen_range(0, 4)) as usize;
+        let num_shards = rng.gen_range(1, 10);
+        let block = block_generator.rand_block(&mut rng, block_size);
+        let block_clone = block.clone();
+        let partitioned = partitioner.partition(block, num_shards);
+        crate::verify_partitioner_output(&block_clone, &partitioned);
+    }
+}
+
+#[test]
+fn test_partitioner_v2_determinism() {
+    let partitioner = Arc::new(PartitionerV2::new(4, 4, 10, 64));
+    assert_deterministic_result(partitioner);
+}
+
+/// Evenly divide 0..n-1. Example: uniform_partition(11,3) == [[0,1,2,3],[4,5,6,7],[8,9,10]]
```

Comment:
The example is actually incorrect, as per your code, we would have chunk sizes of 3,4,4. Am I missing anything?

Reply:
i double checked, the comment is consistent with the test cases.
I'm letting the bigger chunks go first.